### PR TITLE
[acceptance-tests] Add scenario to validate etcd binding and use of detectBindingResources

### PR DESCRIPTION
**Pull request by shvasude**
### Motivation

https://issues.redhat.com/browse/APPSVC-644

### Changes

bind etcd cluster instance
- Given Etcd operator running
-  Given Etcd cluster "etcd-cluster-example" is created
-  Given Imported Nodejs application "nodejs-todo-etcd" is running

        
### Testing

make test-acceptance

For further more details refer the CONTRIBUTING.md

### CODE REVIEW

File: test/acceptance/features/bindAppToMultipleServices.feature [hunk 1]
```diff
@@ -48,7 +48,60 @@ Feature: Bind a single application to multiple services
...
+            | intermediate_secret_name |
+            | binding-request-1        |
+            | binding-request-2        |
+        And jq ".status.conditions[] | select(.type=="CollectionReady").status" of Service Binding Request "binding-request-1" should be changed to "True"
+        And jq ".status.conditions[] | select(.type=="InjectionReady").status" of Service Binding Request "binding-request-1" should be changed to "True"
+        And jq ".status.conditions[] | select(.type=="CollectionReady").status" of Service Binding Request "binding-request-2" should be changed to "True"
+        And jq ".status.conditions[] | select(.type=="InjectionReady").status" of Service Binding Request "binding-request-2" should be changed to "True"
+
+
+    Scenario: Bind two backend services by creating 2 SBRs to a single application
```
Comment (pratikjagrut) [inline test/acceptance/features/bindAppToMultipleServices.feature:61]: This scenario is already present, why are you adding the same scenario? If you wanna change something please change the step only.

File: docs/binding-path.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
+# Custom Containers Path and Secret Path
```
Comment (pedjak) [inline docs/binding-path.md:1]: this file exist on master and should not be part of this PR. Please rebase your PR from master, do not merge master into your branch.

File: test/acceptance/features/bindAppToService.feature [hunk 3]
```diff
@@ -217,3 +217,32 @@ Feature: Bind an application to a service
...
+              backingServiceSelector:
+                group: etcd.database.coreos.com
+                version: v1beta2
+                kind: EtcdCluster
+                resourceRef: etcd-cluster-example
+              detectBindingResources: true
+            """
+        Then deployment must contain intermediate secret
+            | intermediate_secret_name |
+            | binding-request-etcd     |
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:245]: from user perspective, this in unimportant - instead user wants that NodeJS application works. We should check that instead.

File: test/acceptance/features/bindKnativeService.feature [hunk 4]
```diff
@@ -39,7 +39,10 @@ Feature: Bind knative service to a service
                 - name: DB_PASSWORD
                   value: "{{ .knav.status.dbCredentials.password }}"
       """
-    Then deployment must contain intermediate secret "binding-request-knative"
+    Then deployment must contain intermediate secret
+      | intermediate_secret_name |
+      | binding-request-knative  |
```
Comment (pedjak) [inline test/acceptance/features/bindKnativeService.feature:44]: why this change here?

File: test/acceptance/features/steps/steps.py [hunk 5]
```diff
@@ -278,10 +281,11 @@ def then_envFrom_contains(context, app_name, sbr_name1, sbr_name2):
 
 
 # STEP
-@then(u'deployment must contain intermediate secret "{intermediate_secret_name}"')
-def then_envFrom_contains_intermediate_secret_name(context, intermediate_secret_name):
-    assert context.application.get_deployment_with_intermediate_secret(
-        intermediate_secret_name, wait=True, timeout=120) is not None, f"There is no deployment with intermediate secret {intermediate_secret_name}"
+@then(u'deployment must contain intermediate secret')
+def then_envFrom_contains_intermediate_secret_name(context):
```
Comment (pedjak) [inline test/acceptance/features/steps/steps.py:285]: remove `then`

File: test/acceptance/features/steps/steps.py [hunk 6]
```diff
@@ -309,3 +313,29 @@ def verify_injected_secretRef(context, secret_ref, cr_name, crd_name, json_path)
     openshift = Openshift()
     result = openshift.get_resource_info_by_jsonpath(crd_name, cr_name, context.namespace.name, json_path, wait=True, timeout=180)
     result | should.be_equal_to(secret_ref).desc(f'Failed to inject secretRef "{secret_ref}" in "{cr_name}" at path "{json_path}"')
+
+
+@given(u'Etcd operator running')
+def given_etcd_operator_is_running(context):
```
Comment (pedjak) [inline test/acceptance/features/steps/steps.py:319]: remove`given`

File: test/acceptance/features/steps/steps.py [hunk 7]
```diff
@@ -309,3 +313,29 @@ def verify_injected_secretRef(context, secret_ref, cr_name, crd_name, json_path)
...
+    etcd_operator = EtcdOperator("etcd", context.namespace.name)
+    if not etcd_operator.is_running():
+        print("Etcd operator is not installed, installing...")
+        assert etcd_operator.install_operator_subscription() is True, "etcd operator subscription is not installed"
+        assert etcd_operator.is_running(wait=True) is True, "etcd operator not installed"
+    context.etcd_operator = etcd_operator
+
+
+@given(u'Etcd cluster "{etcd_name}" is created')
+def given_etc_cluster_is_created(context, etcd_name):
```
Comment (pedjak) [inline test/acceptance/features/steps/steps.py:332]: remove `given`

File: test/acceptance/features/bindAppToService.feature [hunk 8]
```diff
@@ -217,3 +217,32 @@ Feature: Bind an application to a service
         And jq ".status.conditions[] | select(.type=="InjectionReady").status" of Service Binding Request "binding-request-empty-app" should be changed to "False"
         And jq ".status.conditions[] | select(.type=="InjectionReady").reason" of Service Binding Request "binding-request-empty-app" should be changed to "EmptyApplicationSelector"
 
+    Scenario: Bind an imported Node.js application to Etcd database
+        Given Etcd operator running
+        Given Etcd cluster "etcd-cluster-example" is created
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:222]: rename to `Etcd cluster "etcd-cluster-example" is running`

File: test/acceptance/features/bindAppToMultipleServices.feature [hunk 9]
```diff
@@ -48,7 +48,60 @@ Feature: Bind a single application to multiple services
                     kind: Database
                     resourceRef: db-demo-2
             """
-        Then "nodejs-app" deployment must contain SBR name "binding-request-1" and "binding-request-2"
+        Then deployment must contain intermediate secret
```
Comment (pedjak) [inline test/acceptance/features/bindAppToMultipleServices.feature:51]: can we be more precise what does it? what does it "contain"? in which part? what "intermediate" does mean to users?

**Review by pedjak — CHANGES_REQUESTED** [review]

File: test/acceptance/features/steps/steps.py (hunk 5 above)
Comment (shvasude) [reply test/acceptance/features/steps/steps.py]: done

File: test/acceptance/features/steps/steps.py (hunk 6 above)
Comment (shvasude) [reply test/acceptance/features/steps/steps.py]: done

File: test/acceptance/features/steps/steps.py (hunk 7 above)
Comment (shvasude) [reply test/acceptance/features/steps/steps.py]: done

File: test/acceptance/features/bindAppToService.feature (hunk 8 above)
Comment (shvasude) [reply test/acceptance/features/bindAppToService.feature]: done

File: test/acceptance/features/bindKnativeService.feature (hunk 4 above)
Comment (shvasude) [reply test/acceptance/features/bindKnativeService.feature:44]: I have created a reusable step implementation for `Then deployment must contain intermediate secret` hence i changed here too. I am aware of the PR being related to etcd. but because of this [https://github.com/redhat-developer/service-binding-operator/pull/644#issuecomment-682049475](url) i thought to put it here.

File: test/acceptance/features/bindAppToMultipleServices.feature (hunk 1 above)
Comment (shvasude) [reply test/acceptance/features/bindAppToMultipleServices.feature]: fixed

File: test/acceptance/features/bindKnativeService.feature (hunk 4 above)
Comment (pedjak) [reply test/acceptance/features/bindKnativeService.feature:44]: I would keep the old step where only one secret is mentioned and introduce if really needed additional step, e.g.

```
deployment must contain secrets X, Y, Z
```

btw, what is the meaning of "intermediate secreat" and why a user should care?

File: test/acceptance/features/bindAppToService.feature (hunk 3 above)
Comment (shvasude) [reply test/acceptance/features/bindAppToService.feature:244]: added this because of [https://github.com/redhat-developer/service-binding-operator/pull/567#discussion_r472015323](url)

Can add this step too, `Then application should be connected to the DB "etcd-cluster-example"`, but then we have to generalize this step as `Then application should be connected tot the backing service instance`. your thoughts on this.

File: test/acceptance/features/bindKnativeService.feature (hunk 4 above)
Comment (shvasude) [reply test/acceptance/features/bindKnativeService.feature:44]: added this because of https://github.com/redhat-developer/service-binding-operator/pull/567#discussion_r472015323

File: test/acceptance/features/bindAppToService.feature [hunk 10]
```diff
@@ -216,4 +216,31 @@ Feature: Bind an application to a service
         Then jq ".status.conditions[] | select(.type=="CollectionReady").status" of Service Binding Request "binding-request-empty-app" should be changed to "True"
         And jq ".status.conditions[] | select(.type=="InjectionReady").status" of Service Binding Request "binding-request-empty-app" should be changed to "False"
         And jq ".status.conditions[] | select(.type=="InjectionReady").reason" of Service Binding Request "binding-request-empty-app" should be changed to "EmptyApplicationSelector"
-
+    Scenario: Bind an imported Node.js application to Etcd database
+        Given Etcd operator running
+        * Etcd cluster "etcd-cluster-example" is running
+        * Imported Nodejs application "nodejs-rest-http-crud-etcd" is running
```
Comment (akashshinde) [inline test/acceptance/features/bindAppToService.feature:222]: @shvasude This nodes application will not work with etcd database since it uses postgresql database. Please follow https://github.com/redhat-developer/service-binding-operator/tree/master/examples/nodejs_etcd_operator

**Review by akashshinde — CHANGES_REQUESTED** [review]

File: test/acceptance/features/bindAppToService.feature (hunk 10 above)
Comment (shvasude) [reply test/acceptance/features/bindAppToService.feature:222]: @akashshinde I tried with this nodejs and received this `apiVersion: apps.openshift.io/v1alpha1
kind: ServiceBindingRequest
metadata:
  name: binding-request-etcd
  namespace: test-namespace-b2c50eab
spec:
  applicationSelector:
    group: apps
    version: v1
    resource: deployments
    resourceRef: node-todo-git-app
  backingServiceSelector:
    group: etcd.database.coreos.com
    version: v1beta2
    kind: EtcdCluster
    resourceRef: etcd-cluster-example
  detectBindingResources: true`

Hence used another nodejs application and works everything except validating api end point.

Comment (pedjak) [reply test/acceptance/features/bindAppToService.feature:222]: so how this work if validating endpoint does not work?

File: test/acceptance/features/bindAppToService.feature (hunk 3 above)
Comment (pedjak) [reply test/acceptance/features/bindAppToService.feature:244]: if you can check that the application is working properly, then you do not need this step here - because implicitly we are going to check that. We are not really interested in binding secrets - we are interested that bindings are injected into application so that application works.

File: test/acceptance/features/bindKnativeService.feature (hunk 4 above)
Comment (pedjak) [reply test/acceptance/features/bindKnativeService.feature:44]: not following.

File: test/acceptance/features/bindAppToService.feature (hunk 10 above)
Comment (shvasude) [reply test/acceptance/features/bindAppToService.feature:222]: I thought it works if injectionReady and collectionReady are True and nodejs app fruit app results True for both the status

File: test/acceptance/features/bindKnativeService.feature (hunk 4 above)
Comment (shvasude) [reply test/acceptance/features/bindKnativeService.feature:44]: well, step [](url)(https://github.com/redhat-developer/service-binding-operator/blob/b15ec71a3aa146227a7e40e7e798a260eafb3449/test/acceptance/features/bindAppToMultipleServices.feature#L51)
and [](url)(https://github.com/redhat-developer/service-binding-operator/blob/b15ec71a3aa146227a7e40e7e798a260eafb3449/test/acceptance/features/bindKnativeService.feature#L42) 
can make a similar implementation.what if we need validate more than 2 secrets for binding. we don't need to add another function. hence while implementing [https://github.com/shvasude/service-binding-operator/blob/04673a8e8eebbd6e0a763d2867c18b33ed687331/test/acceptance/features/bindAppToService.feature#L219](url)
added this bdd feature which can be reusable across number of secrets.  
[](url)(https://github.com/shvasude/service-binding-operator/blob/04673a8e8eebbd6e0a763d2867c18b33ed687331/test/acceptance/features/bindAppToService.feature#L242)

File: test/acceptance/features/bindAppToMultipleServices.feature (hunk 9 above)
Comment (shvasude) [reply test/acceptance/features/bindAppToMultipleServices.feature:51]: Can be like this 
`Then "nodejs-app" deployment must contain intermediate secret`

File: test/acceptance/features/bindAppToService.feature (hunk 3 above)
Comment (shvasude) [reply test/acceptance/features/bindAppToService.feature:305]: This commit fixes the use of nodejs todo app instead of nodejs fruit app. 
https://github.com/shvasude/service-binding-operator/commit/e0d93ef37eeb73a61859e5e6220a063a7d4ea67c

File: test/acceptance/features/bindAppToService.feature (hunk 10 above)
Comment (shvasude) [reply test/acceptance/features/bindAppToService.feature]: This commit fixes the use of nodejs todo app instead of nodejs fruit app. 
https://github.com/shvasude/service-binding-operator/commit/e0d93ef37eeb73a61859e5e6220a063a7d4ea67c

File: test/acceptance/features/steps/steps.py [hunk 11]
```diff
@@ -104,7 +104,10 @@ def given_db_operator_is_installed(context):
 @when(imported_nodejs_app_is_running_step)
 def imported_nodejs_app_is_running(context, application_name):
     namespace = context.namespace
-    application = NodeJSApp(application_name, namespace.name)
+    if application_name == 'nodejs-todo-git':
```
Comment (pmacik) [inline test/acceptance/features/steps/steps.py:107]: This does not make much sense, why do you change a behavior of the based on the name? If the Node.js application requires a non-default image, that image should be parametrized in the step sentense itself and not in the implementation based on the name.

Comment (shvasude) [reply test/acceptance/features/steps/steps.py:107]: step sentense? could you please tell me what it means.

Comment (shvasude) [reply test/acceptance/features/steps/steps.py:107]: you mean here in feature file? `Imported Nodejs application "node-todo-git" is running`

File: test/acceptance/features/bindAppToService.feature [hunk 12]
```diff
@@ -307,3 +307,30 @@ Feature: Bind an application to a service
...
+              services:
+                - group: etcd.database.coreos.com
+                  version: v1beta2
+                  kind: EtcdCluster
+                  name: etcd-cluster-example
+              detectBindingResources: true
+            """
+        Then jq ".status.conditions[] | select(.type=="CollectionReady").status" of Service Binding "binding-request-etcd" should be changed to "True"
+        And jq ".status.conditions[] | select(.type=="InjectionReady").status" of Service Binding "binding-request-etcd" should be changed to "True"
+        And deployment must contain intermediate secret "binding-request-etcd"
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:336]: Application endpoint step got introduced in this PR: `https://github.com/redhat-developer/service-binding-operator/pull/613/files#diff-ec891afbdb0cf5fd075cbf5ca535a45fR118 but it is not used anywhere. IMHO, instread of checking that deployment must contain the secret, checking if application endpoint works is actually ultimate check and we should apply that instead.

Comment (pmacik) [reply test/acceptance/features/bindAppToService.feature]: Good point, I've updated and squashed...

**Comment by pmacik** [comment]
I've rebased on top of most recent changes to `master` (merged master, fixed conflicts and squashed)

**Review by pedjak — APPROVED** [review]
